W1405: inconsistent-quotes does not respect py-version

### Bug description

Since Python 3.12, the same quotes can be used in f-strings as were used for the outer string itself. Pylint correctly points out that inconsistent quote styles have been used if this is not done. However, it does not take into account the configured py-version, so it also warns for Python versions < 3.12 when the test is performed with Python 3.12.

```python
dictionary = {'0': 0}
f_string = f'{dictionary["0"]}'
```


### Configuration

```ini
py-version=3.10
```


### Command used

```shell
pylint a.py
```


### Pylint output

```python
************* Module a
a.py:2:0: W1405: Quote delimiter " is inconsistent with the rest of the file (inconsistent-quotes)
```


### Expected behavior

As far as I understand it, pylint should only give this warning if the py-version is >= 3.12.

### Pylint version

```shell
pylint 3.0.0
astroid 3.0.0
Python 3.12.0 (main, Oct  3 2023, 22:08:15) [GCC 13.2.1 20230801]
```

--- a/doc/whatsnew/fragments/9113.bugfix
+++ b/doc/whatsnew/fragments/9113.bugfix
@@ -1,0 +1,3 @@
+Emit ``inconsistent-quotes`` for f-strings with 3.12 interpreter only if targeting pre-3.12 versions.
+
+Closes #9113

--- a/pylint/checkers/strings.py
+++ b/pylint/checkers/strings.py
@@ -841,6 +841,7 @@ class StringConstantChecker(BaseTokenChecker, BaseRawFileChecker):
         # First, figure out which quote character predominates in the module
         for tok_type, token, _, _, _ in tokens:
             if sys.version_info[:2] >= (3, 12):
+                # pylint: disable=no-member,useless-suppression
                 if tok_type == tokenize.FSTRING_START:
                     inside_fstring = True
                 elif tok_type == tokenize.FSTRING_END:

--- a/pylint/checkers/strings.py
+++ b/pylint/checkers/strings.py
@@ -19,6 +19,7 @@ from astroid.typing import SuccessfulInferenceResult
 
 from pylint.checkers import BaseChecker, BaseRawFileChecker, BaseTokenChecker, utils
 from pylint.checkers.utils import only_required_for_messages
+from pylint.constants import PY312_PLUS
 from pylint.interfaces import HIGH
 from pylint.typing import MessageDefinitionTuple
 
@@ -834,8 +835,21 @@ class StringConstantChecker(BaseTokenChecker, BaseRawFileChecker):
         """
         string_delimiters: Counter[str] = collections.Counter()
 
+        inside_fstring = False  # whether token is inside f-string (since 3.12)
+        target_py312 = self.linter.config.py_version >= (3, 12)
+
         # First, figure out which quote character predominates in the module
         for tok_type, token, _, _, _ in tokens:
+            if PY312_PLUS:
+                if tok_type == tokenize.FSTRING_START:
+                    inside_fstring = True
+                elif tok_type == tokenize.FSTRING_END:
+                    inside_fstring = False
+
+                if inside_fstring and not target_py312:
+                    # skip analysis of f-string contents
+                    continue
+
             if tok_type == tokenize.STRING and _is_quote_delimiter_chosen_freely(token):
                 string_delimiters[_get_quote_delimiter(token)] += 1
 

--- a/pylint/checkers/strings.py
+++ b/pylint/checkers/strings.py
@@ -8,6 +8,7 @@ from __future__ import annotations
 
 import collections
 import re
+import sys
 import tokenize
 from collections import Counter
 from collections.abc import Iterable, Sequence
@@ -19,7 +20,6 @@ from astroid.typing import SuccessfulInferenceResult
 
 from pylint.checkers import BaseChecker, BaseRawFileChecker, BaseTokenChecker, utils
 from pylint.checkers.utils import only_required_for_messages
-from pylint.constants import PY312_PLUS
 from pylint.interfaces import HIGH
 from pylint.typing import MessageDefinitionTuple
 
@@ -840,7 +840,7 @@ class StringConstantChecker(BaseTokenChecker, BaseRawFileChecker):
 
         # First, figure out which quote character predominates in the module
         for tok_type, token, _, _, _ in tokens:
-            if PY312_PLUS:
+            if sys.version_info[:2] >= (3, 12):
                 if tok_type == tokenize.FSTRING_START:
                     inside_fstring = True
                 elif tok_type == tokenize.FSTRING_END:

--- a/tests/functional/i/inconsistent/inconsistent_quotes_fstring.py
+++ b/tests/functional/i/inconsistent/inconsistent_quotes_fstring.py
@@ -1,0 +1,4 @@
+# pylint: disable=missing-module-docstring
+
+dictionary = {'0': 0}
+f_string = f'{dictionary["0"]}'

--- a/tests/functional/i/inconsistent/inconsistent_quotes_fstring.rc
+++ b/tests/functional/i/inconsistent/inconsistent_quotes_fstring.rc
@@ -1,0 +1,5 @@
+[STRING]
+check-quote-consistency=yes
+
+[testoptions]
+max_pyver=3.12

--- a/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312.py
+++ b/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312.py
@@ -1,0 +1,5 @@
+# pylint: disable=missing-module-docstring
+
+dictionary = {'0': 0}
+# quotes are inconsistent when targetting Python 3.12 (use single quotes)
+f_string = f'{dictionary["0"]}'  # [inconsistent-quotes]

--- a/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312.rc
+++ b/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312.rc
@@ -1,0 +1,8 @@
+[STRING]
+check-quote-consistency=yes
+
+[main]
+py-version=3.12
+
+[testoptions]
+min_pyver=3.12

--- a/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312.txt
+++ b/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312.txt
@@ -1,0 +1,1 @@
+inconsistent-quotes:5:0:None:None::"Quote delimiter "" is inconsistent with the rest of the file":UNDEFINED

--- a/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312_311.py
+++ b/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312_311.py
@@ -1,0 +1,5 @@
+# pylint: disable=missing-module-docstring
+
+dictionary = {'0': 0}
+# quotes are consistent when targetting 3.11 and earlier (cannot use single quotes here)
+f_string = f'{dictionary["0"]}'

--- a/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312_311.rc
+++ b/tests/functional/i/inconsistent/inconsistent_quotes_fstring_py312_311.rc
@@ -1,0 +1,8 @@
+[STRING]
+check-quote-consistency=yes
+
+[main]
+py-version=3.11
+
+[testoptions]
+min_pyver=3.12

